File: Backend/app/models/alert.py
from sqlalchemy import Column, Integer, String, Boolean, DateTime, ForeignKey, Text, Enum
from sqlalchemy.orm import relationship
from datetime import datetime
import enum
from app.db.session import Base

class AlertType(str, enum.Enum):
    HORS_ZONE = "HORS_ZONE"
    VITESSE_EXCESSIVE = "VITESSE_EXCESSIVE"
    ARRET_PROLONGE = "ARRET_PROLONGE"
    MOTEUR_COUPE = "MOTEUR_COUPE"
    BATTERIE_FAIBLE = "BATTERIE_FAIBLE"

class AlertSeverity(str, enum.Enum):
    FAIBLE = "FAIBLE"
    MOYENNE = "MOYENNE"
    CRITIQUE = "CRITIQUE"

class Alert(Base):
    __tablename__ = "alerte"

    id_alerte = Column(Integer, primary_key=True, index=True)
    id_vehicule = Column(Integer, ForeignKey("vehicule.id_vehicule"), nullable=False)
    id_position = Column(Integer, ForeignKey("position_gps.id_position"))
    type_alerte = Column(String(50), nullable=False, index=True)
    severite = Column(String(20), default=AlertSeverity.MOYENNE.value)
    message = Column(Text, nullable=False)
    details_json = Column(Text)
    action_prise = Column(String(100))
    acquittee = Column(Boolean, default=False, index=True)
    acquittee_par = Column(Integer, ForeignKey("utilisateur.id_utilisateur"))
    acquittee_at = Column(DateTime)
    created_at = Column(DateTime, default=datetime.utcnow, index=True)

    # Relationships
    vehicule = relationship("Vehicle", back_populates="alertes")

File: Frontend/lib/models/alert.dart
enum AlertType {
  horsZone,
  vitesseExcessive,
  arretProlonge,
  moteurCoupe,
  batterieFaible,
}

enum AlertSeverity { faible, moyenne, critique }

class Alert {
  final String id;
  final String vehicleId;
  final AlertType type;
  final AlertSeverity severity;
  final String message;
  final DateTime timestamp;
  final bool isAcknowledged;

  Alert({
    required this.id,
    required this.vehicleId,
    required this.type,
    this.severity = AlertSeverity.moyenne,
    required this.message,
    required this.timestamp,
    this.isAcknowledged = false,
  });

  Alert copyWith({
    String? id,
    String? vehicleId,
    AlertType? type,
    AlertSeverity? severity,
    String? message,
    DateTime? timestamp,
    bool? isAcknowledged,
  }) {
    return Alert(
      id: id ?? this.id,
      vehicleId: vehicleId ?? this.vehicleId,
      type: type ?? this.type,
      severity: severity ?? this.severity,
      message: message ?? this.message,
      timestamp: timestamp ?? this.timestamp,
      isAcknowledged: isAcknowledged ?? this.isAcknowledged,
    );
  }
}
